ui/selectors/selectors.js

```diff
@@ -536,15 +536,41 @@ export const getInternalAccountsSortedByKeyring = createDeepEqualSelector(
   getMetaMaskKeyrings,
   getMetaMaskAccounts,
   (keyrings, accounts) => {
+    const thirdpartySnaps = 'thirdpartySnaps';
+    // Create a map of entropySource map to accounts for quick lookup
+    const entropySourceToAccountsMap = Object.values(accounts).reduce(
+      (map, account) => {
+        if (account.metadata?.keyring?.type === KeyringTypes.snap) {
+          const { entropySource = thirdpartySnaps } = account.options || {};
+          if (!map[entropySource]) {
+            map[entropySource] = [];
+          }
+          map[entropySource].push(account);
+        }
+        return map;
+      },
+      {},
+    );
+
     // keep existing keyring order
-    const internalAccounts = keyrings
-      .map(({ accounts: addresses }) => addresses)
-      .flat()
-      .map((address) => {
-        return accounts[address];
-      });
+    return keyrings.reduce((internalAccounts, keyring) => {
+      // Get regular accounts for this keyring
+      const keyringAccounts = keyring.accounts.map(
+        (address) => accounts[address],
+      );
 
-    return internalAccounts;
+      // If it's an HD keyring, add any snap accounts that belong to it
+      if (keyring.type === KeyringTypes.hd) {
+        const snapAccounts =
+          entropySourceToAccountsMap[keyring.metadata.id] || [];
+        return [...internalAccounts, ...keyringAccounts, ...snapAccounts];
+      } else if (keyring.type === KeyringTypes.snap) {
+        const thirdpartySnapAccounts =
+          entropySourceToAccountsMap[thirdpartySnaps] || [];
+        return [...internalAccounts, ...thirdpartySnapAccounts];
+      }
+      return [...internalAccounts, ...keyringAccounts];
```

Comment:
I think it is a bit faster if you just push the new elements into internalAccounts instead of generating a new array for every iteration

```suggestion
      // If it's an HD keyring, add any snap accounts that belong to it
      if (keyring.type === KeyringTypes.hd) {
        const snapAccounts =
          entropySourceToAccountsMap[keyring.metadata.id] || [];
        internalAccounts.push(...keyringAccounts, ...snapAccounts);
        return internalAccounts;
      } else if (keyring.type === KeyringTypes.snap) {
        const thirdpartySnapAccounts =
          entropySourceToAccountsMap[thirdpartySnaps] || [];
        internalAccounts.push(...thirdpartySnapAccounts);
        return internalAccounts;
      }
      internalAccounts.push(...keyringAccounts);
      return internalAccounts;
```